Offchain indexing example & writeup (#435)
JoshOrndorff/recipes: text/off-chain-workers/indexing.md

# Off-chain Indexing

There are times when on-chain extrinsics need to pass data to the off-chain worker context, such as pass the extrinsic caller account to off-chain worker for further computation.

> We can surely pass this piece of data via on-chain storage, but this is costly and it will make the data propagate among the blockchain network. If this is not a piece of critical information that need to have consensus upon the whole network, another way is to pass this data via off-chain indexing.

Reviewer: Can you think of a real-world motivating example? I guess there must be one if this feature was added to Substrate.

Author: @tomusdrw Could you give a real-world example when an on-chain context need to pass data to the off-chain worker context? Or why did you implement this feature in the first place?

Reviewer 2: Like I mentioned earlier, the main use case for `offchain_index` is it's predictable write behavior - for all imported blocks it is guaranteed to run.

Examples in our current codebase include:
1. Writing historic validator sets to offchain db for eqiuvocation reporting purposes (we only keep hash on-chain, so we can verify a proof).
2. Writing leaves & nodes of Merkle Mountain Range, while only keeping peaks (and hashes of leaves) on-chain.

We will demonstrate this in [`ocw-demo` pallet](https://github.com/substrate-developer-hub/recipes/tree/master/pallets/ocw-demo/src/lib.rs).
Knowledge discussed in this chapter built upon [using local storage in off-chain worker context](./storage.md).

> **Notes**
>
> In order to see the off-chain indexing feature in effect, please run the kitchen node with

> > off-chain indexing flag on, as `./target/debug/kitchen-node --dev --tmp --enable-offchain-indexing true`

Reviewer: You've used the debug path here. Is that intentional? I just realized the recipes no longer explains how to run a node.

}
```

We first define a key used in the local off-chain storage. Then we can write to the storage with
`offchain_index::set(key, value)` function. Here `offchain_index::set()` accepts values in byte
format (`&[u8]`) so we encode the number first. If you refer back to
[`offchain_index` API rustdoc](https://substrate.dev/rustdocs/v3.0.0/sp_io/offchain_index/index.html),
you will see there are only `set()` and `clear()` functions. This means from the on-chain context,
we only expect to write to this local off-chain storage location but not reading from it, and we
cannot pass data within on-chain context using this method.

## Reading the Data in Off-chain Context

src: [`pallets/ocw-demo/src/lib.rs`](https://github.com/substrate-developer-hub/recipes/tree/master/pallets/ocw-demo/src/lib.rs)

```rust
fn offchain_worker(block_number: T::BlockNumber) {
  // -- snip --

  // Reading the number written in the last on-chain transaction.
  let mem_onchain_num = StorageValueRef::persistent(ONCHAIN_TX_KEY);
  if let Some(Some(onchain_num)) = mem_onchain_num.get::<u64>() {
    debug::info!("Number written on last on-chain transaction: {:?}", onchain_num);
  }
}
```

> We read the data back in the `offchain_worker()` function by how we would normally read from the

Reviewer: ```suggestion
We read the data back in the `offchain_worker()` function as we would normally read from the
```

# Off-chain Indexing

There are times when on-chain extrinsics need to pass data to the off-chain worker context with
predictable write behavior. We can surely pass this piece of data via on-chain storage, but this is
costly and it will make the data propagate among the blockchain network. If this is not a piece of
information that need to have consensus upon the whole network, another way is to save this data
in off-chain local storage via off-chain indexing.

As off-chain indexing is called in on-chain context, **if it is agreed upon by the blockchain
consensus mechanism, then it is expected to run predictably by all nodes in the network**. One use case
is to store only the hash of certain information in on-chain storage for verification purpose but
keeping the full data set off-chain for lookup later. In this case the original data can be saved
via off-chain indexing.

Notice as off-chain indexing is called and data is saved on block import, the result may be

> overridden should the block IS NOT FINALIZED eventually, or a fork appears in the blockchain.

Author: @tomusdrw I have updated this part based on your comment. See if they look good now. Thanks.

# Off-chain Indexing

There are times when on-chain extrinsics need to pass data to the off-chain worker context with
predictable write behavior. We can surely pass this piece of data via on-chain storage, but this is
costly and it will make the data propagate among the blockchain network. If this is not a piece of

> information that need to have consensus upon the whole network, another way is to save this data

Reviewer 2: ```suggestion
information that need to have consensus upon the whole network, another way is to save this data
```

indexing is a piece of information we've reach consensus about, it's just not polluting the on-chain state.

in off-chain local storage via off-chain indexing.

As off-chain indexing is called in on-chain context, **if it is agreed upon by the blockchain
consensus mechanism, then it is expected to run predictably by all nodes in the network**. One use case
is to store only the hash of certain information in on-chain storage for verification purpose but
keeping the full data set off-chain for lookup later. In this case the original data can be saved
via off-chain indexing.

> Notice as off-chain indexing is called and data is saved on block import, the result may be
> overridden should the block IS NOT FINALIZED eventually, or a fork appears in the blockchain.

Reviewer 2: ```suggestion
Notice as off-chain indexing is called and data is saved on every block import, this also includes forks.
The consequence of that is that in case non-unique keys are used the data might be overwritten by different
forked blocks and the content of off-chain database will be different between nodes.
Care should be taken when choosing the right indexing `key`, to prevent potential overwrites if it's not desired.
```

Author: updated in the latest commit
